### PR TITLE
MolGAN Porting

### PR DESCRIPTION
**Pull request by shreyasvinaya**
## Description
Porting of MolGAN from TF to torch

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [ ] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [ ] Run `mypy -p deepchem` and check no errors
  - [ ] Run `flake8 <modified file> --count` and check no errors
  - [ ] Run `python -m doctest <modified file>` and check no errors
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New unit tests pass locally with my changes
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/models/torch_models/molgan.py [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+class BasicMolGANModel(WGANModel):
+    """
+    Model for de-novo generation of small molecules based on work of Nicola De Cao et al. [1]_.
+    It uses a GAN directly on graph data and a reinforcement learning objective to induce the network to generate molecules with certain chemical properties.
+    Utilizes WGAN infrastructure; uses adjacency matrix and node features as inputs.
+    Inputs need to be one-hot representation.
+
+    Examples
+    --------
+    >>>
```
Comment (rbharath) [inline deepchem/models/torch_models/molgan.py:20]: This won't run unit test

File: deepchem/models/torch_models/molgan.py [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+
+    def create_discriminator(self,
+                             units: List[Union[Tuple[int, int],
+                                               int]] = [(128, 64), 64]):
+        """
+        Create discriminator model based on MolGAN layers.
+        Takes two inputs:
+            1. adjacency tensor, containing bond information
+            2. nodes tensor, containing atom information
+        The input vectors need to be in one-hot encoding format.
```
Comment (rbharath) [inline deepchem/models/torch_models/molgan.py:148]: Missing parameter fields in docs

File: deepchem/models/torch_models/molgan.py [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+                             units=units,
+                             edges=self.edges)
+
+    def predict_gan_generator(self,
+                              batch_size: int = 1,
+                              noise_input: Optional[List] = None,
+                              conditional_inputs: List = [],
+                              generator_index: int = 0) -> List[GraphMatrix]:
+        """
+        Use the GAN to generate a batch of samples.
```
Comment (rbharath) [inline deepchem/models/torch_models/molgan.py:213]: Add usage example

File: deepchem/models/torch_models/molgan.py [hunk 4]
```diff
@@ -0,0 +1,398 @@
...
+
+
+class BasicMolGANGenerator(nn.Module):
+    """
+    Generator class for BasicMolGAN model.
+    Using subclassing rather than functional API due to requirement
+    to swap between two outputs depending on situation.
+    In order to get output that used for sample generation
+    (conversion to rdkit molecules) pass sample_generation=True argument while
+    calling the model i.e. adjacency_matrix, nodes_features = self.generators[0](
```
Comment (rbharath) [inline deepchem/models/torch_models/molgan.py:262]: Add usage example

File: deepchem/models/torch_models/tests/test_molgan_model.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+            if valid_molecules_count:
+                success = True
+                break
+
+        # finally test if there was at least one valid training session
+        # as the model structure improves this should become more and more strict
+        assert success
+
+
+if __name__ == '__main__':
```
Comment (rbharath) [inline deepchem/models/torch_models/tests/test_molgan_model.py:143]: Remove before final version

**Comment by shreyasvinaya** [comment]
Moved to https://github.com/deepchem/deepchem/pull/3814

**Pull request closed without merging**